pmda/util.py

```diff
@@ -30,3 +57,118 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+    remainder block).
+
+    Arguments
+    ---------
+    n_frames : int
+        number of frames in the trajectory (≥0). This must be the number of
+        frames *after* the trajectory has been sliced,
+        i.e. ``len(u.trajectory[start:stop:step])`` and `start` and `step` must
+        be provided as parameters.
+    n_blocks : int
+        number of blocks (>0)
+    start : int
+        first index of the trajectory (default is ``None``, which is
+        interpreted as "first frame", i.e., 0); see comments on `n_frames`.
+    step : int
+        step by which the trajectory is sliced (see description of `n_frames`);
+        the default is ``None`` which corresponds to ``step=1``.
+
+    Returns
+    -------
+    frame_indices : array
+        Array of shape of length ``n_blocks + 1`` with the starting frame of
+        each block; the last index ``frame_indices[-1]`` corresponds to the
+        last index of the last block + 1 so that one can easily use the array
+        for slicing, as shown in the example below.
+
+        If `n_frames` = 0 then an empty list ``[]`` is returned.
+
+        .. note:: For `step` > 1 the last index in `frame_indices` may be
+                  larger than the real last index in the trajectory; this is
+                  not a problem for slicing but it means that one should *not*
+                  rely on the last index for any other operations except
+                  slicing of the original trajectory.
+
+    Example
+    -------
+    For a trajectory with 5 frames and 4 blocks we get block sizes ``[2, 1, 1,
+    1]`` (instead of ``[2, 2, 1, 0]``).
+
+    The indices will be ``[0, 2, 3, 4, 5]``.
+
+    The indices can be used to slice a trajectory into blocks::
+
+        n_blocks = 5
+        n_frames = len(u.trajectory[start:stop:step])
+
+        idx = assign_blocks(n_frames, n_blocks)
+        for i_block, (bstart, bstop) in enumerate(zip(idx[:-1], idx[1:])):
+           for ts in u.trajectory[bstart:bstop:step]:
+               # do stuff for block number i_block
+
+    Note that in order to access the correct frames in each block, the
+    trajectory *must be sliced with the original `step` value* (``step`` in the
+    example) and the start and stop indices returned by
+    :func:`make_balanced_blocks`, namely ``bstart`` and ``bstop`` in the
+    example above.
+
+
+    Notes
+    -----
+    Explanation of the algorithm: For `N` trajectory frames in the trajectory
+    and `N` blocks (or processes), where `i` with 0 ≤ `i` ≤ `N` - 1 is
+    the block number and `m[i]` is the number of frames for block `i` we
+    get a *balanced distribution* (one that does not contain blocks of size 0)
+    with the algorithm ::
+
+        m[i] = M // N     # initial frames for block i
+        r = M % N         # remaining frames 0 ≤ r < N
+        for i in range(r):
+            m[i] += 1     # distribute the remaining frames
+                          # over the first r blocks
```

Comment:
why spreading it out? We could just add the remainder in a new block. It would get rid of this loop.

Reply:
Then you don't have N blocks anymore. That leads to bad load balancing if you asked for N blocks because you have N cores and now we suddenly give you N+1 blocks.

Furthermore, as you can see in the code, the actual implementation with numpy arrays is simpler than that...

This algorithm gives you a better distribution of frames per block than what we had before. It also leads to a maximum difference of 1 between blocks – just by construction. It guarantees that you get exactly as many blocks as you asked for. I don't really think one can do better.

Reply:
Even if you don't use `n_blocks == n_jobs`, it is still better to have frames distributed across blocks as evenly as possible. I think under normal circumstances adding a number of frames on the order of the number of blocks to the last block is not going to make a big difference because the number of frames tends to be much larger than the number of blocks – but it is always going to be worse than distributing the load evenly because this one task will have a different run time from the others. Dask with heavily oversubscribed workers might load balance this (even though IIRC that did not seem a great strategy for trajectory analysis) but MPI would prefer even load. 

Again, it might be a small effect compared to many other things but why use an inferior algorithm when the better one is equally cheap? 

> We only have our balancing problem because we set `n_blocks == n_jobs`

We don't enforce this – the user if free to choose `n_blocks`. Only if `n_blocks == None` will we choose what is likely the best strategy, namely as many blocks as workers.

And the original problem #71 was simply that we could end up with blocks with 0 frames. That needed to be fixed one way or another. My point is that it is better to distribute frames as evenly as possible over `n_blocks` (where the number of frames maximally differs by 1 between any two blocks) than to do an uneven distribution where the last block can have up to `n_blocks - 1` frames more than all the other blocks.